pe/packrat.py

```diff
@@ -298,15 +298,16 @@ def _debug(self, definition: Definition) -> _Matcher:
         def _match(s: str, pos: int, memo: Memo) -> RawMatch:
             # for proper printing, only terminals can print after
             # knowing the result
+            snippet = f"{repr(s[pos:pos+10])[1:-1][:12]:<12}"
```

Comment:
I like the idea of pre-computing the snippet instead of doing string manipulation in each branch of the conditional below, but I'm not fond of using `repr()` and then cleaning up the quotes. You can use `.encode("unicode_escape")`, although that returns a bytestring so it then needs to be decoded to ASCII ([link](https://stackoverflow.com/a/14961000/1441112)):

```python
snippet = s.encode("unicode_escape").decode("ascii")[:12].ljust(12)
```

Reply:
Yeah that looks nicer, have updated the PR.

Reply:
@TomHodson Sorry, I introduced a bug such that it only takes the first 12 characters of the input. It also inefficiently encodes the entire input string. It should be something like this:

```python
snippet = s[pos:pos+10].encode("unicode_escape").decode("ascii")[:10].ljust(12)
```

There are a few other issues with non-latin unicode characters (e.g., Japanese `あ` becomes `\u3042`, which is 6 characters), but I think I can deal with these myself after merging.

Reply:
I agree that characters that encode to many bytes will be an issue. Perhaps we should only replace newlines and carriage returns? That way the line is still readable but you avoid expanding non-latin characters and emoji. 
```
s = """
😊
あ\r
"""
pos = 0
snippet = s[pos:pos+10].replace("\n", r"\n").replace("\r", r"\r")[:10].ljust(12)
print(snippet)
```
gives "\n😊\nあ\r\n"

Reply:
Yes I was thinking of something like that. There are several line-breaking spaces:
* `\n` newline
* `\r` carriage return
* `\v` vertical tab
* `\f` form feed
* NEL next line (U+0085)
* line separator (U+2028)
* paragraph separator (U+2029)

In addition, things like literal tab characters can disrupt the display of the debug context, so we might replace those.

And if we output literal double-width characters like `あ`, then it will cause ragged edges in the debug context. We could use ANSI escapes to move the position to an absolute position before printing the `|` for the divider between the context and grammar rules. This would only display properly in a terminal or somewhere that interprets ANSI sequences, though, so if debug output were piped to a file, it would not look good.

Reply:
For the escaping, how about some thing like the folllowing. The dictionary seems a bit redundant but as there are multiple ways to escape each character (eg "\r", "\x...". "\u...", "\N{...}"), it seems like it gives the easiest fine grained control.

```python
def escape_line_breaking_characters(text : str) -> str:
    """Escape any characters that create large amounts of whitespace,
    mainly these are line breaking characters like newlines and also tabs.
    """
    offending_characters = {
        "\n" : r"\n", # newline
        "\r" : r"\r", # carriage return
        "\v" : r"\v", # vertical tab
        "\f" : r"\f", # form feed
       "\t"  : r"\t" # tab
        "\u0085" : r"\u0085", # NEL next line
        "\u2028" : r"\u2028", # \N{LINE SEPARATOR}
        "\u2029" : r"\u2029", # \N{PARAGRAPH SEPARATOR}
    }
    return "".join(offending_characters.get(c, c) for c in text)
    

test = "A\nB\r\nC\vD\fE\u0085F\u2028G\u2029H"
print(test)
print()
print(escape_line_breaking_characters(test))
```
prints
```
A
B
CDEF G H

A\nB\r\nC\vD\fE\u0085F\u2028G\u2029H
```

I have no ideas for the double width characters, if we're not using a monospace font then is there any hope of getting the pipe characters to align anyway? I generally use jupyter notebooks to evaluate snippets of python code so I have a vested interest in the solution working in places where ANSI sequences are not obeyed.

Reply:
And perhaps nice would be to use str.maketrans and str.translate:
```python
def escape_line_breaking_characters(text : str) -> str:
    """Escape any characters that create large amounts of whitespace,
    mainly these are line breaking characters like newlines and also tabs.
    """
    offending_characters = str.maketrans({
        "\n" : r"\n", # newline
        "\r" : r"\r", # carriage return
        "\v" : r"\v", # vertical tab
        "\f" : r"\f", # form feed
        "\u0085" : r"\u0085", # NEL next line
        "\u2028" : r"\u2028", # \N{LINE SEPARATOR}
        "\u2029" : r"\u2029", # \N{PARAGRAPH SEPARATOR}
    })
    return text.translate(offending_characters)
```

Reply:
> I generally use jupyter notebooks to evaluate snippets of python code so I have a vested interest in the solution working in places where ANSI sequences are not obeyed.

Ah good point. Ok, I agree that ANSI escapes are not a good general solution here.

> use str.maketrans and str.translate

That is a great idea. I'd forgotten about that feature. My go-to in these situations is `re.sub()` with a function as the replacement, but the native string method is a better fit.

And I think your proposed function (maybe adding `'\t': r'\t',  # horizontal tab`), along with the snippet extraction, would be sufficient to close this ticket.

I'm prepared to live with ragged edges for double-width characters. We _could_ detect whether the unicode codepoint is double-width and adjust accordingly, no ANSI escapes required, but I don't think it's worth the effort at this time. What I would rather do, eventually, is have the debug operation take a callback function so it can be configured. E.g., for Jupyter output we could possibly print an HTML table or something. But that's too big a change for now.